File: python/p018.py
#!/usr/bin/python


def calc_sum_at_row(nums, row_num):
    for i in range(len(nums[row_num])):
        nums[row_num][i] += max([nums[row_num + 1][i], nums[row_num + 1][i + 1]])

    if len(nums[row_num]) == 1:
        return nums[row_num][0]
    else:
        return calc_sum_at_row(nums, row_num - 1)


nums = []
with open("../data/p018.txt") as f:
    nums = f.read().strip(' \t\n\r"').split('\n')

for i in range(len(nums)):
    nums[i] = [int(i) for i in nums[i].split(' ')]

print(calc_sum_at_row(nums, len(nums) - 2))

File: rust/src/bin/p018.rs

use std::io::Read;
use std::fs::File;

fn calc_sum_at_row(mut nums: Vec<Vec<usize>>, row_num: usize) -> usize {

    for i in (0..nums[row_num].len()) {
        nums[row_num][i] += *[nums[row_num + 1][i], nums[row_num + 1][i + 1]].iter().max().unwrap();
    }

    if nums[row_num].len() == 1 {
        return nums[row_num][0];
    } else {
        return calc_sum_at_row(nums.clone(), row_num - 1);
    }
}

fn main() {

    let mut s = String::new();
    let mut f = File::open("../data/p018.txt").unwrap();
    let _ = f.read_to_string(&mut s);

    let mut nums: Vec<Vec<usize>> = vec![];

    for l in s.trim_matches('\n').split("\n") {
        nums.push(l.split(" ").map(|x| x.parse::<usize>().unwrap()).collect());
    }

    let row_num = nums.len() - 2;
    println!("{}", calc_sum_at_row(nums, row_num));

}
